modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -1110,14 +1110,22 @@ def _maybe_reorder_labels(
             row_order_mapping = dict(
                 zip(sorted_row_positions, range(len(row_positions)))
             )
-            new_row_order = [row_order_mapping[idx] for idx in row_positions]
+            new_row_order = np.fromiter(
+                (row_order_mapping[idx] for idx in row_positions),
+                dtype="int64",
+                count=len(row_positions),
+            )
```

Comment:
from the surrounding code, I believe we can make the following:
1. `sorted_row_positions` is a numpy array (we can make it 100% sure by calling `numpy.asarray()` which is a no-op for a thing which is already an array)
2. same for `row_positions`
3. then making the dictionary and then building an array via `np.fromiter` is just a super-slow version of `sorted_row_positions[row_positions]` 😉 

If I'm right, this piece could be sped up even further (and code greatly simplified) by rewriting as
```python
new_row_order = np.asarray(sorted_row_positions)[np.asarray(row_positions)]
```

Reply:
What does it matter if it's a contiguous array (if you mean in-memory contiguous)? Numpy is smart enough:
```python
>>> np.array([1,5,7,2,9,12,19,-5,6])[np.array([2,6])]
array([ 7, 19])
>>> x1=np.array(np.arange(10))
>>> x1.flags
  C_CONTIGUOUS : True
  F_CONTIGUOUS : True
  OWNDATA : True
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

>>> x2=x1[::-1]
>>> x2.flags
  C_CONTIGUOUS : False
  F_CONTIGUOUS : False
  OWNDATA : False
  WRITEABLE : True
  ALIGNED : True
  WRITEBACKIFCOPY : False

>>> x2[np.array([2,3])]
array([7, 6])
>>> x1[x2]
array([9, 8, 7, 6, 5, 4, 3, 2, 1, 0])
```

Reply:
I have the following example in my mind:
```python
>>> import numpy as np
>>> row_positions = np.array([7,1,3])
>>> sorted_row_positions = np.array([1,3,7])
>>> sorted_row_positions[row_positions]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
IndexError: index 7 is out of bounds for axis 0 with size 3
>>>
```

Reply:
I have built an alternative approach, need to benchmark it before making a PR though...